### DISCUSSION TITLE
1 pixel difference in Y axis

### DISCUSSION THREAD

**Discussion by gastendonk** [red6/pdfcompare#124] [Q&A] [answered]
We have switched from Java 8 to 17. Now there's a difference of (I guess) 1 pixel in Y axis. Is there already a solution for this to ignore this difference?

![image](https://user-images.githubusercontent.com/31237274/178027762-cbe829a0-b043-4e52-8983-d51a8542a495.png)

jasperreports:6.4.3

**Comment by gastendonk** [comment, marked as answer]
This is the cause of the problem and there will be no fix for it: https://bugs.openjdk.org/browse/JDK-8215290

So that the transition from Java 8 to Java 17 still works, I modified the jasperPrint object before the PDF file is created and pdfcompare is executed.

The leadingOffset values of the JRTemplatePrintText class differ for a JasperReport created with Java 8 and one with Java 17.

    private void fake(JasperPrint jasperPrint) {
        for (JRPrintPage page : jasperPrint.getPages()) {
            fake2(page.getElements());
        }
    }
    private void fake2(List<JRPrintElement> list) {
        for (JRPrintElement e : list) {
            if (e instanceof JRTemplatePrintText) {
                JRTemplatePrintText text = (JRTemplatePrintText) e;
                // The values have to be adjusted.
                if (text.getLeadingOffset() == -1.9072266f) {
                    text.setLeadingOffset(-1.7600098f);
                } else if (text.getLeadingOffset() == -1.6953125f) {
                    text.setLeadingOffset(-1.5644531f);
                } else if (text.getLeadingOffset() == -2.5429688f) {
                    text.setLeadingOffset(-2.3466797f);
                } else if (text.getLeadingOffset() == -2.1191406f) {
                    text.setLeadingOffset(-1.9555664f);
                } else if (text.getLeadingOffset() != 0f) {
                    System.out.println("<> " + text.getLeadingOffset());
                }
            } else if (e instanceof JRPrintFrame) {
                fake2(((JRPrintFrame) e).getElements()); // recursive
            }
        }
    }

**Reply by finsterwalder** [reply]
This is not a problem in PdfCompare, is it? It's more a problem that you compare PDF documents generated with different Java versions? So your system, after upgrading Java produces slightly different documents due to slight changes in font metrics, right?

**Comment by gastendonk** [comment]
formula for getting the new value: text.getLeadingOffset() / 1.08364544334f
